From 70c66fb67b6aa5b1ef401191b8879552d52d081f Mon Sep 17 00:00:00 2001
From: Shengjiu Wang <b02247@freescale.com>
Date: Thu, 3 Apr 2014 16:05:40 +0800
Subject: [PATCH 0183/1594] ENGR00306857 pulseaudio5.0 mute Headphone volume
 when Headphone plugged

commit 6035d35f5ea6e590d9c3eaa9aa7b9b95abb2bb1f from
git://git.freescale.com/imx/linux-2.6-imx.git

Pulseaudio will detect the Headphone Jack, then swith to Headphone.
So register new Jack for Headphone, the iface=CARD.

Signed-off-by: Shengjiu Wang <b02247@freescale.com>
(cherry picked from commit 6a715373c43f16e48883061049e67919281878d1)
---
 sound/soc/fsl/Kconfig      |    1 +
 sound/soc/fsl/imx-wm8962.c |   37 ++++++++++++++++++++++++++-----------
 2 files changed, 27 insertions(+), 11 deletions(-)

diff --git a/sound/soc/fsl/Kconfig b/sound/soc/fsl/Kconfig
index 19c302b..7429558 100644
--- a/sound/soc/fsl/Kconfig
+++ b/sound/soc/fsl/Kconfig
@@ -236,6 +236,7 @@ config SND_SOC_IMX_WM8962
 	select SND_SOC_IMX_PCM_DMA
 	select SND_SOC_IMX_AUDMUX
 	select SND_SOC_FSL_SSI
+	select SND_KCTL_JACK
 	help
 	  Say Y if you want to add support for SoC audio on an i.MX board with
 	  a wm8962 codec.
diff --git a/sound/soc/fsl/imx-wm8962.c b/sound/soc/fsl/imx-wm8962.c
index 60ee03a..2532240 100644
--- a/sound/soc/fsl/imx-wm8962.c
+++ b/sound/soc/fsl/imx-wm8962.c
@@ -22,6 +22,7 @@
 #include <linux/clk.h>
 #include <sound/soc.h>
 #include <sound/jack.h>
+#include <sound/control.h>
 #include <sound/pcm_params.h>
 #include <sound/soc-dapm.h>
 #include <linux/pinctrl/consumer.h>
@@ -50,6 +51,8 @@ struct imx_priv {
 	struct platform_device *pdev;
 	struct snd_pcm_substream *first_stream;
 	struct snd_pcm_substream *second_stream;
+	struct snd_kcontrol *headphone_kctl;
+	struct snd_card *snd_card;
 };
 static struct imx_priv card_priv;
 
@@ -81,7 +84,7 @@ static struct snd_soc_jack_gpio imx_mic_jack_gpio = {
 	.invert = 0,
 };
 
-static int hpjack_status_check(void)
+static int hpjack_status_check(void *data)
 {
 	struct imx_priv *priv = &card_priv;
 	struct platform_device *pdev = priv->pdev;
@@ -103,10 +106,12 @@ static int hpjack_status_check(void)
 		snprintf(buf, 32, "STATE=%d", 2);
 		snd_soc_dapm_disable_pin(&priv->codec->dapm, "Ext Spk");
 		ret = imx_hp_jack_gpio.report;
+		snd_kctl_jack_report(priv->snd_card, priv->headphone_kctl, 1);
 	} else {
 		snprintf(buf, 32, "STATE=%d", 0);
 		snd_soc_dapm_enable_pin(&priv->codec->dapm, "Ext Spk");
 		ret = 0;
+		snd_kctl_jack_report(priv->snd_card, priv->headphone_kctl, 0);
 	}
 
 	envp[0] = "NAME=headphone";
@@ -118,7 +123,7 @@ static int hpjack_status_check(void)
 	return ret;
 }
 
-static int micjack_status_check(void)
+static int micjack_status_check(void *data)
 {
 	struct imx_priv *priv = &card_priv;
 	struct platform_device *pdev = priv->pdev;
@@ -278,9 +283,10 @@ static struct snd_soc_ops imx_hifi_ops = {
 	.hw_free = imx_hifi_hw_free,
 };
 
-static int imx_wm8962_gpio_init(struct snd_soc_pcm_runtime *rtd)
+static int imx_wm8962_gpio_init(struct snd_soc_card *card)
 {
-	struct snd_soc_codec *codec = rtd->codec;
+	struct snd_soc_dai *codec_dai = card->rtd[0].codec_dai;
+	struct snd_soc_codec *codec = codec_dai->codec;
 	struct imx_priv *priv = &card_priv;
 
 	priv->codec = codec;
@@ -289,9 +295,10 @@ static int imx_wm8962_gpio_init(struct snd_soc_pcm_runtime *rtd)
 		imx_hp_jack_gpio.gpio = priv->hp_gpio;
 		imx_hp_jack_gpio.jack_status_check = hpjack_status_check;
 
-		snd_soc_jack_new(codec, "Headphone Jack", SND_JACK_HEADPHONE, &imx_hp_jack);
-		snd_soc_jack_add_pins(&imx_hp_jack,
-				ARRAY_SIZE(imx_hp_jack_pins), imx_hp_jack_pins);
+		snd_soc_card_jack_new(card, "Headphone Jack",
+						SND_JACK_HEADPHONE, &imx_hp_jack,
+						imx_hp_jack_pins,
+						ARRAY_SIZE(imx_hp_jack_pins));
 		snd_soc_jack_add_gpios(&imx_hp_jack, 1, &imx_hp_jack_gpio);
 	}
 
@@ -299,9 +306,10 @@ static int imx_wm8962_gpio_init(struct snd_soc_pcm_runtime *rtd)
 		imx_mic_jack_gpio.gpio = priv->mic_gpio;
 		imx_mic_jack_gpio.jack_status_check = micjack_status_check;
 
-		snd_soc_jack_new(codec, "AMIC", SND_JACK_MICROPHONE, &imx_mic_jack);
-		snd_soc_jack_add_pins(&imx_mic_jack,
-				ARRAY_SIZE(imx_mic_jack_pins), imx_mic_jack_pins);
+		snd_soc_card_jack_new(card, "AMIC",
+					SND_JACK_MICROPHONE, &imx_mic_jack,
+					imx_mic_jack_pins,
+					ARRAY_SIZE(imx_mic_jack_pins));
 		snd_soc_jack_add_gpios(&imx_mic_jack, 1, &imx_mic_jack_gpio);
 	} else if (priv->amic_mono || priv->dmic_mono) {
 		/*
@@ -481,7 +489,6 @@ static int imx_wm8962_probe(struct platform_device *pdev)
 	data->dai.cpu_dai_name = dev_name(&ssi_pdev->dev);
 	data->dai.platform_of_node = ssi_np;
 	data->dai.ops = &imx_hifi_ops;
-	data->dai.init = &imx_wm8962_gpio_init;
 	data->dai.dai_fmt = SND_SOC_DAIFMT_I2S | SND_SOC_DAIFMT_NB_NF |
 			    SND_SOC_DAIFMT_CBM_CFM;
 
@@ -509,6 +516,14 @@ static int imx_wm8962_probe(struct platform_device *pdev)
 		goto fail;
 	}
 
+	priv->snd_card = data->card.snd_card;
+	priv->headphone_kctl = snd_kctl_jack_new("Headphone", 0, NULL);
+	ret = snd_ctl_add(data->card.snd_card, priv->headphone_kctl);
+	if (ret)
+		goto fail;
+
+	imx_wm8962_gpio_init(&data->card);
+
 	if (gpio_is_valid(priv->hp_gpio)) {
 		ret = driver_create_file(pdev->dev.driver, &driver_attr_headphone);
 		if (ret) {
-- 
1.7.5.4

